src/ompl/geometric/planners/bitstar/src/BITstar.cpp

```diff
@@ -914,11 +909,32 @@ namespace ompl
             //A new start pointer
             const ompl::base::State* newStart;
 
+            //If the list of goal vertices is empty, wait until the first goal state is available.
+            if (goalVertices_.empty())
+            {
+                newGoal = Planner::pis_.nextGoal(ptc);
+                if (newGoal)
+                {
+                    //It is valid, allocate the vertex pointer
+                    goalVertices_.push_back(boost::make_shared<Vertex>(Planner::si_, opt_));
+
+                    //Copy the value into the state
+                    Planner::si_->copyState(goalVertices_.back()->state(), newGoal);
+
+                    //And add this goal to the set of samples:
+                    this->addSample(goalVertices_.back());
+
+                    //Mark that we've added:
+                    addedState = true;
+                }
+            }
+
             //Add the new starts and goals to the lists of said vertices.
             //Do goals first, as they are only added as samples.
-            //If newGoal is NULL, then no new goal exists. The following line both assigns newGoal and checks if it is a valid pointer.
-            while (static_cast<bool>((newGoal = Planner::pis_.nextGoal(ptc))) == true)
+            while (pis_.haveMoreGoalStates())
```

Comment:
Ok.
So (right now) `addStartsAndGoalStates(ptc)` gets called in two places:

1. at the end of `setup()` with `ompl::base::plannerAlwaysTerminatingCondition`, and
2. in `newBatch(ptc)` with `ptc`. 

`newBatch(ptc)` itself is only called in two places:

1. **if** there are no goals, at the start of `solve(ptc)` before `iterate()` is called. It is given `ptc` with the intention it will wait for a goal, and
2. every time we finish the batch in `iterate` with `ompl::base::plannerAlwaysTerminatingCondition`.

As it's written now, `addStartsAndGoalStates(ptc)` always calls `PlannerInputStates::nextGoal(ptc)` and then checks whether the returned pointer is `NULL` or not. How the call to `nextGoal(ptc)` behaves depends on the `ptc`

1. If it is `ompl::base::plannerAlwaysTerminatingCondition`, `nextGoal(ptc)` gets any available goals, but if there are none does not wait for more to appear. It will return NULL in that case. This is what we want to happen once we're up and running with at least 1 goal.
2. If it is a real `PlannerTerminatingCondition`, `nextGoal(ptc)` will return any available goals and if there are none, wait for them to appear as per the `ptc`. This is what we want to happen in the case that `solve(ptc)` has been called without a goal available.
  1. Ah ha! Here **is** a problem, it will **always** use all of the `ptc` waiting for goals, won't it?... Is this what you were trying to fix? I will have to take a closer look at this.

P.S.,
In explaining this I realised it should be a call to `addStartsAndGoalStates(ptc)` at the top of `solve(ptc)`, instead `newBatch(ptc)`. As it's purpose is just to repeat the call at the end of `setup()`.

Also that it should be `addStartAndGoalsStates(ptc)` and not addStart**s**AndGoalsStates(ptc)